README.md

```diff
@@ -33,106 +33,131 @@ A project was created with those packages `npm install`ed, creating a gigantic `
-
-### `import` Statements of CommonJS Packages and Files
+File extensions are still relevant. While either a `.js` or an `.mjs` file can be loaded as ESM, only `.js` files can be loaded as CommonJS. If the above example’s `cookies.js` was renamed `cookies.mjs`, the theoretical `import cookies from 'request/lib/cookies.mjs'` would throw.
 
-Because of the problem of potentially importing the same package or identifier as both ESM or CommonJS, and therefore creating two module map entries for the same identifier, we need to be careful when we allow CommonJS to be imported into ESM files. Currently, this proposal allows importing of CommonJS only via `createRequireFromPath`; `import` statements beginning with a CommonJS package name; or `import` statements of an absolute URL where a `package.json` at or above the file being imported creates a CommonJS package boundary around the file. Other than these exceptions, `import` statements always import JavaScript as ESM. If issues are discovered with allowing `import` statements of CommonJS files (whether as deep imports or via an absolute URL), support for those cases will be dropped and users will need to use `createRequireFromPath` or any other methods we may create in the future. The priority is to enable `import` statements of CommonJS package main entry points; importing of CommonJS files inside or outside of packages will be supported whenever possible.
+The CommonJS automatic file extension resolution or folder `index.js` discovery are not supported for `import` statements, even when referencing files inside CommonJS packages. Both `import cookies from 'request/lib/cookies'` and `import request from './node_modules/request'` would throw. Automatic file extension resolution or folder `index.js` discovery _are_ still supported for `package.json` `"main"` field specifiers, however, to preserve backward compatibility.
 
-#### `import` statements of CommonJS package root/main entry point
+### Specifiers starting with `/` or `//`
 
-An ESM JavaScript file importing a CommonJS package, where the entire import specifier is the CommonJS package name, uses the specifier resolution algorithm that CommonJS uses now to resolve the main entry point of the package. For example, where `underscore` and `pad` are both CommonJS packages:
+These are currently unsupported but reserved for future use. Browsers support specifiers like `'/app.js'` to be relative to the base URL of the page, whereas in CommonJS a specifier starting with `/` refers to the root of the file system. We would like to find a solution that conforms Node closer to browsers for `/`-leading specifiers.
```

Comment:
I'd be weary of defining "/" as the package root as it is not a clearly browser compatible pattern at all as there is no way browsers could ever implement this.

I agree this is a worthy use case, but it should be done at least with en eye to browser compatibility.

Reply:
@robpalme Mapping `/` to package root was the original draft of this proposal (you can look in the git history) but we removed it because we weren’t sure of the implications of that for browser compatibility, hence the current language.

I think there’s pretty much no reason to map `/` to file system root, as that would generally only work on the author’s system; and if you still really wanted it, you could still achieve it using a `file:///` URL. So the question would be how to provide a mapping to package root, whether via `/` or `~` or something else, that’s compatible (or at least not a conflict) with browsers.

Reply:
If we had `import:` and the idea that packages are domains means that `import:/…` is the same as `import://this-package/` (ie package root), and `import://that-package/`... And assuming the `import:` scheme defaults to `//` as the prefix of a bare pathname then `import:/` is the same also the same as `import:this-package/` just like `import:that-package` and `import://that-package/`

Reply:
Yes exactly, there are many ways to achieve this one!

The package exports proposal could even be extended to support it - `exports: { "mypkg": "./" }`.

Reply:
Absolutely, and I think that people looking for that convenience to connotational are mostly those who have relied on tooling that handled that while transpiling their source (or special/hybrid platforms) so it is really outside the scope of this proposal.

That said, do you think it at all justified to at least try hinting at this it is with the a statement along the lines of [my suggestion above](#r236077665)?

Reply:
Seems good to me, it's nice to mention this one and that we are compatible with various approaches here.

Reply:
I'd just note that `/` already has significant meaning to URLs in the `file:` scheme. It goes to the `file:` URL's base namespace which could be a host:

```terminal
> new URL('/', 'file:///app').href
'file:///'
> new URL('//', 'file:///app').href
'file:///'
> new URL('/', 'file://laptop/app').href
'file://laptop/'
> new URL('//', 'file://desktop/app').href
'file://desktop/'
```

Per relevant tests you can see https://github.com/web-platform-tests/wpt/blob/4ba3cd24c69a859d8c7864d15958d2656ef16173/url/resources/urltestdata.json#L5539

Windows drive letters have some special behavior and drop the host though due to what appears to be legacy issues.